README.md
# Riemann test project 

## just testing Rieman with Alerta

*   Disable riemann with RabbitMQ
    In the file `riemann_config/riemann.confg`
    comment the line `(include "/etc/riemann/conf.d")`

### Invoke Riemann

1. Invoke
    ```
    docker-compose up -d riemann
    ```
2. Understand what happening
   ```
    docker logs riemann
   ```


### Invoke Alerta

   In testing environment, we assume the host machine that running docker is using IP `10.20.30.40`.

   ```
   docker-compose up -d alerta-web
   ```
   Use web browser to see http://10.20.30.40:8081


### Setup input testing strems

*   Use riemann-tools to collect cpu, memory, disk, load and send metric to Riemann 
    ```
    gem install riemann-tools
    riemann-health --host my.riemann.server
    ```

## Testing Riemann with RabbitMQ

*   Enable riemann with RabbitMQ
    In the file `riemann_config/riemann.confg`
    uncomment the line `; (include "/etc/riemann/conf.d")`

### The rabbitMQ exchange needs to be setup by publisher

1.  Invoke mq first
   ```
    docker-compose up -d mq
   ```
2.  Invoke mq publisher
   ```
    cd mqclient
    lein run
   ```
3.  Invoke riemann
   ```
    docker-compose up -d riemann
   ```


mqclient/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-03-07
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-03-07
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/mqclient/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mqclient/compare/0.1.0...0.1.1


mqclient/test/mqclient/core_test.clj
(ns mqclient.core-test
  (:require [clojure.test :refer :all]
            [mqclient.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


riemann_config/conf.d/test_streams.clj
    ; streams stay in let block of main riemann 
    (streams
        (where (not (state "expired"))
            prn
            (match :service #"disk"
                (with {:event "FsUtil" :group "OS"}
                    (splitp < (* metric 100)
                        95 (critical "Filesystem utilisation >95% " dedup-alert)
                        90 (major "Filesystem utilisation >90%" dedup-alert)
                        80 (warning "Filesystem utilisation >80%" dedup-alert)
                        (normal "Filesystem utilisation is OK" dedup-alert))))))


conf.d/temp.clj
; Expire old events from the index every 5 seconds.
(periodically-expire 5)

(let [index (index)]
  ; Inbound events will be passed to these streams:
  (streams
    (default :ttl 60
      ; Index all events immediately.
      index

      ; show all the events received.
      ; #(info "received event" %)

      ; Log expired events.
      ; (expired
      ;    (fn [event] (info "expired" event))))))

      ; select certain service
      (where (service "net.if.in.bits/iface=eth_all")
        ; split streams by host and alarm on per host condition
        (by :host
          (fixed-time-window 5
            (smap folds/median
              (where (< metric 80.0)
                #(info "alert event!" %)))))

        ; split steams by platform and alarm on aggregated metrics.
        (where (= (:platform event) "c01.i01")
          (fixed-time-window  5
            (smap folds/sum
              (where (> metric 100.0)
                #(info "alert platform event!!!" %)))))))))


streams_src/main.go
package main

import (
	"github.com/amir/raidman"
	"time"
)

func main() {
	c, err := raidman.Dial("tcp", "10.20.30.40:5555")
	if err != nil {
		panic(err)
	}

	// send multiple events at once
	for i := 0; i < 10; i++ {
		err = c.SendMulti([]*raidman.Event{
			&raidman.Event{
				State:      "success",
				Host:       "ctl-00",
				Service:    "net.if.in.bits/iface=eth_all",
				Metric:     10 * i,
				Ttl:        10,
				Attributes: map[string]string{"platform": "c01.i01"},
			},
			&raidman.Event{
				State:      "failure",
				Host:       "ctl-01",
				Service:    "net.if.in.bits/iface=eth_all",
				Metric:     10 * (i + 1),
				Ttl:        10,
				Attributes: map[string]string{"platform": "c01.i01"},
			},
			&raidman.Event{
				State:      "success",
				Host:       "ctl-02",
				Service:    "net.if.in.bits/iface=eth_all",
				Metric:     10 * (i + 2),
				Ttl:        10,
				Attributes: map[string]string{"platform": "c01.p02"},
			},
		})
		if err != nil {
			panic(err)
		}
		time.Sleep(5000 * time.Millisecond)
	}

	events, err := c.Query("host = \"ctl-00\"")
	if err != nil {
		panic(err)
	}

	if len(events) < 1 {
		panic("Submitted event not found")
	}

	c.Close()
}


riemann_config/alerta.clj
; -*- mode: clojure; -*-
; vim: filetype=clojure

(def version "3.1.0")

(def alerta-endpoints
    {:alert "http://10.20.30.40:8081/api/alert"
    :heartbeat "http://10.20.30.40:8081/api/heartbeat"})

(defn post-to-alerta
  "POST to the Alerta REST API."
  [url request]
  (let [event-url url
    event-json (json/generate-string request)]
    (client/post event-url
               {:body event-json
                :socket-timeout 5000
                :conn-timeout 5000
                :content-type :json
                :accept :json
        ; :debug true
        ; :debug-body true
                :throw-entire-message? true})))

(defn format-alerta-event
  "Formats an event for Alerta."
  [event]
  {
   :origin (str "riemann/" hostname)
   :resource
    (if (.contains (:service event) " ")
      (let [[_ instance] (clojure.string/split (:service event) #" " 2)]
        (str (:host event) ":" instance))
        (:host event))
   :event (get event :event (:service event))
   :group (get event :group "Performance")
   :value (:metric event)
   :severity (:state event)
   :environment (get event :environment "Production")
   :service [(get event :grid "Platform")]
   :tags (if (empty? (:tags event)) [] (:tags event))
   :text (:description event)
   :rawData event})

(defn alerta
  "Creates an alerta adapter.
    (changed-state (alerta))"
  [e]
  (post-to-alerta (:alert alerta-endpoints) (format-alerta-event e)))

(defn heartbeat [e] (post-to-alerta
    (:heartbeat alerta-endpoints)
    {:origin (str "riemann/" hostname)
       :tags [version]
       :type "Heartbeat"}))



riemann_config/conf.d/mq.clj
(load-plugins) ; will load plugins from the classpath

(rabbitmq-plugin/amqp-consumer {
                       :parser-fn rabbitmq-plugin/logstash-parser ; message parsing function, the sample function here is the default
                       :prefetch-count 100 ; this is the default
                       :bindings [{
                         :opts {:durable true :auto-delete false} ; this is the default
                         :queue "queue.default" ; the default is "" which means auto-generated queue name
                         :bind-to {"ex", ["queue.default"]} ; also works with single non-seq binding key
                         :tags ["amqp"] ; will be added to event tags
                       }]
                       :connection-opts {:host "10.20.30.40" :port 5672 :username "guest" :passowrd "guest"} ; default is {}
                       })


mqclient/src/mqclient/core.clj
(ns mqclient.core
  (:gen-class)
  (:require [langohr.core      :as rmq]
            [langohr.channel   :as lch]
            [langohr.queue     :as lq]
            [langohr.exchange  :as lhe]
            [langohr.basic     :as lb]
            [clojure.data.json :as json]))

(defn message-handler
  [ch {:keys [content-type delivery-tag type] :as meta} ^bytes payload]
  (println (format "[consumer] Received a message: %s, delivery tag: %d, content type: %s, type: %s"
                   (String. payload "UTF-8") delivery-tag content-type type)))

(def c_message
  {:host   "127.0.0.3"
   :service "cpu"
   :metric 0.9 })

(def message
  {:host   "127.0.0.5"
   :service "disk"
   :metric 0.9 })

(def payload
  (json/write-str message))

(defn -main
  [& args]
  (let [conn  (rmq/connect)
        ch    (lch/open conn)
        qname "queue.default"
        routingkey "queue.default"
        exchange "ex"]
    (println (format "[main] Connected. Channel id: %d" (.getChannelNumber ch)))
    (lq/declare ch qname {:durable true :exclusive false :auto-delete false})
    ;
    ; when using default exchange, which name is "".
    ; we do not need to declare exchange or bind exchange with queue.
    ;
    (lhe/declare ch exchange "direct")
    (lq/bind ch qname exchange)
    (lb/publish ch exchange routingkey payload {:content-type "text/plain" :type "greetings.hi"})
    (Thread/sleep 2000)
    (println "[main] Disconnecting...")
    (rmq/close ch)
    (rmq/close conn)))


riemann_config/conf.d/basic_streams.clj
    ; streams stay in let block of main riemann 
    (streams
        (where (not (state "expired"))
            prn
            (match :service "heartbeat"
                (fn [event]
                    (let [elapsed (- (now) (:metric event))
                        ttl (:ttl event)]
                        ((with {:event "Heartbeat" :group "Riemann" :metric elapsed}
                            (splitp < elapsed
                                90 (major "Heartbeat stale by more than 90 seconds" dedup-alert)
                                60 (minor "Heartbeat stale by more than 60 seconds" dedup-alert)
                                30 (warning "Heartbeat stale by more than 30 seconds" dedup-alert)
                                (normal "Hearbeat received within " ttl " seconds and not stale" dedup-alert))) event))))

            (match :service "load"
                (with {:event "SystemLoad" :group "OS"}
                    (splitp < metric
                        0.7 (major "15-min load average is very high" dedup-alert)
                        0.4 (warning "15-min load average is high" dedup-alert)
                        (normal "15-min load average is OK" dedup-alert))))

            (match :service "cpu"
                (with {:event "CpuUtil" :group "OS"}
                    (splitp < (* metric 100)
                        99 (critical "CPU utilisation >99%" dedup-alert)
                        95 (major "CPU utilisation >95%" dedup-alert)
                        90 (minor "CPU utilisation >90%" dedup-alert)
                        80 (warning "CPU utilisation >80%" dedup-alert)
                        (normal "CPU utilisation is OK" dedup-alert))))

            (match :service "memory"
                (with {:event "MemUtil" :group "OS"}
                    (splitp < (* metric 100)
                        90 (major "Memory utilisation >90%" dedup-alert)
                        75 (warning "Memory utilisation >75%" dedup-alert)
                        (normal "Memory utilisation is OK" dedup-alert))))

            (match :service #"inode"
                (with {:event "FsUtil" :group "OS"}
                    (splitp < (* metric 100)
                        95 (critical "Filesystem utilisation >95% " dedup-alert)
                        90 (major "Filesystem utilisation >90%" dedup-alert)
                        80 (warning "Filesystem utilisation >80%" dedup-alert)
                        (normal "Filesystem utilisation is OK" dedup-alert))))))
